### PR TITLE
Fetch protocol version and use versioned remote procedures

### PR DESCRIPTION
**Pull request by soerenreichardt**

### CODE REVIEW

File: graphdatascience/query_runner/aura_db_query_runner.py [hunk 1]
```diff
@@ -50,8 +53,15 @@ def call_procedure(
         if params is None:
             params = CallParameters()
 
-        if AuraDbQueryRunner.GDS_REMOTE_PROJECTION_PROC_NAME == endpoint:
-            return self._remote_projection(endpoint, params, yields, database, logging)
+        if AuraDbQueryRunner.GDS_REMOTE_PROJECTION_PROC_NAME in endpoint:
+            if self._endpoint_has_supported_version(endpoint):
+                return self._remote_projection(endpoint, params, yields, database, logging)
```
Comment (Mats-SX) [inline graphdatascience/query_runner/aura_db_query_runner.py:58]: so in this PR we send the same remote projection query regardless of protocol version?

File: graphdatascience/query_runner/aura_db_query_runner.py [hunk 2]
```diff
@@ -120,32 +130,43 @@ def _remote_write_back(
...
-            "jobId": job_id,
-            "arrowConfiguration": db_arrow_config,
-        }
+        graph_name = params["graph_name"]
+        if ProtocolVersion.V2 in self._protocol_versions:
+            db_write_proc_params = self._write_back_params_v2(graph_name, job_id, db_arrow_config, config)
+            protocol_version = ProtocolVersion.V2
+        elif ProtocolVersion.V1 in self._protocol_versions:
+            db_write_proc_params = self._write_back_params_v1(graph_name, job_id, db_arrow_config)
+            protocol_version = ProtocolVersion.V1
```
Comment (Mats-SX) [inline graphdatascience/query_runner/aura_db_query_runner.py:154]: I was expecting to see something similar to this for projections as well.

**Review by Mats-SX — CHANGES_REQUESTED** [review]
I think we need to think about the error cases and what instruction we should give the user in those cases.

Also, we need to run different queries for the different protocol versions also when projecting, right?

File: graphdatascience/query_runner/aura_db_query_runner.py [hunk 3]
```diff
@@ -50,8 +53,15 @@ def call_procedure(
...
-        if AuraDbQueryRunner.GDS_REMOTE_PROJECTION_PROC_NAME == endpoint:
-            return self._remote_projection(endpoint, params, yields, database, logging)
+        if AuraDbQueryRunner.GDS_REMOTE_PROJECTION_PROC_NAME in endpoint:
+            if self._endpoint_has_supported_version(endpoint):
+                return self._remote_projection(endpoint, params, yields, database, logging)
+            else:
+                raise RuntimeError(
+                    f"Unsupported procedure: {endpoint}. \
+                    The database supports the following procedure protocol versions: {self._protocol_versions}. \
+                    Please use a matching project procedure."
```
[the remark this answers is not in the document]
Comment (soerenreichardt) [reply graphdatascience/query_runner/aura_db_query_runner.py:63]: No you're right. But we should instead point out that the user should update it's client version.

[the remark this answers is not in the document]
Comment (soerenreichardt) [reply graphdatascience/query_runner/aura_db_query_runner.py:63]: I also removed the sentence above. The user should not be informed about any versions, as they are purely internal

File: graphdatascience/query_runner/aura_db_query_runner.py [hunk 4]
```diff
@@ -120,32 +130,43 @@ def _remote_write_back(
         logging: bool = False,
         custom_error: bool = True,
     ) -> DataFrame:
-        if params["config"] is None:
-            params["config"] = {}
```
[the remark this answers is not in the document]
Comment (soerenreichardt) [reply graphdatascience/query_runner/aura_db_query_runner.py:124]: I missed that we pass the whole `params` dict in the procedure call below. Fixed it now.

File: graphdatascience/query_runner/aura_db_query_runner.py (hunk 2 above)
Comment (soerenreichardt) [reply graphdatascience/query_runner/aura_db_query_runner.py]: Yes, you're right I confused a bit which of the projection procedures and aggregation functions is called in which order 🤷

File: graphdatascience/query_runner/aura_db_query_runner.py (hunk 1 above)
Comment (soerenreichardt) [reply graphdatascience/query_runner/aura_db_query_runner.py]: Fixed now

File: graphdatascience/session/aura_graph_data_science.py [hunk 5]
```diff
@@ -194,3 +202,20 @@ def close(self) -> None:
         Close the GraphDataScience object and release any resources held by it.
         """
         self._query_runner.close()
+
+    def _db_protocol_versions(self) -> List[ProtocolVersion]:
```
Comment (Mats-SX) [inline graphdatascience/session/aura_graph_data_science.py:206]: we should test this behaviour

File: graphdatascience/query_runner/aura_db_query_runner.py [hunk 6]
```diff
@@ -123,29 +199,46 @@ def _remote_write_back(
...
-        }
+        self._inject_arrow_config(db_arrow_config)
+
+        graph_name = params["graph_name"]
+        if ProtocolVersion.V2 in self._server_protocol_versions:
+            db_write_proc_params = self._write_back_params_v2(graph_name, job_id, db_arrow_config, config)
+            protocol_version = ProtocolVersion.V2
+        elif ProtocolVersion.V1 in self._server_protocol_versions:
+            db_write_proc_params = self._write_back_params_v1(
+                graph_name, self._gds_query_runner.database(), job_id, db_arrow_config
```
Comment (Mats-SX) [inline graphdatascience/query_runner/aura_db_query_runner.py:224]:
```suggestion
                graph_name, self._db_query_runner.database(), job_id, db_arrow_config
```

**Review by Mats-SX — APPROVED** [review]

**Review by Mats-SX — CHANGES_REQUESTED** [review]
Before we merge, we need to manually integration test with 5.24-dev

**Review by Mats-SX — APPROVED** [review]

File: graphdatascience/tests/integration/test_graph_construct.py [hunk 7]
```diff
@@ -499,8 +499,6 @@ def test_graph_construct_with_arrow_no_db() -> None:
         gds.close()
         pytest.skip("Arrow server not enabled")
 
-    assert not gds.database()
```
Comment (Mats-SX) [inline graphdatascience/tests/integration/test_graph_construct.py:502]: let's revert this

**Pull request merged**